jupyterhub/user.py

```diff
@@ -235,8 +235,11 @@ def spawn(self, options=None):
         api_token = self.new_api_token()
         db.commit()
 
+        self.server_base_url = base_url
+
         spawner = self.spawner
```

Comment:
I don't think Spawner needs to be modified at all, except perhaps for how it accesses `self.server` instead of `self.user.server`. The User just needs to have a dict of Spawner instances instead of a singleton.